[Merged by Bors] - Add `finalized` to HTTP API responses

danielrachi1 opened the pull request:
## Issue Addressed

#3708 

## Proposed Changes
- Add `is_finalized_block` method to `BeaconChain` in `beacon_node/beacon_chain/src/beacon_chain.rs`.
- Add `is_finalized_state` method to `BeaconChain` in `beacon_node/beacon_chain/src/beacon_chain.rs`.
- Add `fork_and_execution_optimistic_and_finalized` in `beacon_node/http_api/src/state_id.rs`.
- Add `ExecutionOptimisticFinalizedForkVersionedResponse` type in `consensus/types/src/fork_versioned_response.rs`.
- Add `execution_optimistic_finalized_fork_versioned_response`function in  `beacon_node/http_api/src/version.rs`.
- Add `ExecutionOptimisticFinalizedResponse` type in `common/eth2/src/types.rs`.
- Add `add_execution_optimistic_finalized` method in  `common/eth2/src/types.rs`.
- Update API response methods to include finalized.
- Remove `execution_optimistic_fork_versioned_response`

danielrachi1 committed:
Add finalized flag to /eth/v2/beacon/blocks/{block_id}

michaelsproul commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -1179,6 +1180,7 @@ pub fn serve<T: BeaconChainTypes>(
              accept_header: Option<api_types::Accept>| {
                 async move {
                     let (block, execution_optimistic) = block_id.full_block(&chain).await?;
+                    let finalized = block_id.is_finalized(&chain);
```
In the cases where we have a `block_id` we also usually have the `block` already loaded, so it's wasteful to load it again.

To accomplish this I think the method on `BeaconChain` should look something like this:

```rust
pub fn is_finalized_block(&self, block_root: &Hash256, block_slot: Slot) -> Result<bool, Error>;
```

The implementation could check that `block_slot` is less than or equal to the finalized slot _and_ that the canonical root at `block_slot` matches `block_root`. The canonical check is necessary because the database may have some non-canonical blocks and states which conflict with finalization.

```rust
let finalized_slot = chain
    .canonical_head
    .cached_head()
    .finalized_checkpoint()
    .epoch
    .start_slot(E::slots_per_epoch());

let is_canonical = chain
    .block_root_at_slot(block_slot, WhenSlotSkipped::None)?
    .map_or(false, |canonical_root| block_root == canonical_root);

block_slot <= finalized_slot && is_canonical
```

For states something similar is required (check the slot _and_ the canonical status).

If we want to optimize these a little we can also return an optional finality status from a method like `BlockId::root_and_finality_status` returning `Result<(Hash256, Option<bool>), Error>`

danielrachi1 replied:
I'm confused here. My current understanding is:
As the Beacon-API docs say, `block_id`s can be one of: "head", "genesis", "finalized", \<slot\>, \<hex encoded blockRoot with 0x prefix\>. Those variants are represented in the `CoreBlockId` enum that's wrapped into the `BlockId` tuple struct.

So, when someone calls a path that includes a `block_id` we'd either have a slot _or_ a root.

For this reason, I don't understand how can I implement a function that takes both the slot and the root of the block I want to check. If I start with the slot, I'd have to call `chain.block_root_at_slot(slot,...)`, to get a root I can pass to the function. This would beat the purpose of passing both the slot and the root because the root would always be canonical. If I start with the root, I'd have to call `chain.get_blinded_block(root)` to get a slot, which again, would beat the purpose of checking if it's canonical because the slot would always be calculated from the root.

danielrachi1 replied:
Also, in what sense is the block already loaded when having the `block_id`?

michaelsproul replied:
For the `GET beacon/blocks/{block_id}` API I mean we've already loaded the block on the previous line:

```rust
let (block, execution_optimistic) = block_id.full_block(&chain).await?;
```

So I was imagining checking that it's finalized like so:

```rust
let block_root = block_id.root(&chain)?;
let finalized = chain.is_finalized_block(block_root, block.slot())?;
```

This avoids repeating an expensive database read loading the block from the database a second time.

> If I start with the slot, I'd have to call chain.block_root_at_slot(slot,...), to get a root I can pass to the function.

This observation is correct: when we do lookups by slot we almost always return the canonical block, so checking that it's canonical in this case is redundant (we just need to check if it's prior to the finalized slot). Structuring the code to allow this would be a nice optimisation, but not 100% necessary.

> If I start with the root, I'd have to call chain.get_blinded_block(root) to get a slot, which again, would beat the purpose of checking if it's canonical because the slot would always be calculated from the root.

This isn't quite right though: there's only 1 canonical block root per slot, but multiple block roots _may_ point to blocks with the same slot. So we do still need to check whether the root is canonical (but _do not_ need to load the block more than once).

Some form of the `block` is loaded by all the APIs which accept a `{block_id}`, so similar logic can be applied for all of them.

I hope that makes sense, let me know if anything is still unclear :blush:

danielrachi1 replied:
That made everything clear, thank you! Pls see my latest commit to check if I missed something :)

Also, should I delete my original `is_finalized()`? I think it still could be useful in case we don't have the `block` loaded.

And one minor thing: I think the name `is_block_finalized` is a bit cleaner than `is_finalized_block`, is there a reason you choose that name specifically? haha

michaelsproul replied:
Latest commit looks great!

> Also, should I delete my original is_finalized()? I think it still could be useful in case we don't have the block loaded.

I think we may as well delete it, as far as I can tell all the APIs have the block already loaded, even the `/root` API.

> And one minor thing: I think the name is_block_finalized is a bit cleaner than is_finalized_block, is there a reason you choose that name specifically?

I don't mind either way, I chose `is_finalized_block` because it's quite common to use finalized as a prefix adjective like "download a finalized block", etc. But `is_block_finalized` is fine too

michaelsproul commented on beacon_node/http_api/src/block_id.rs:
```diff
@@ -211,6 +211,38 @@ impl BlockId {
+                    .unwrap()
+                    .slot()
+                    .epoch(T::EthSpec::slots_per_epoch());
+                block_epoch <= finalized_epoch
```
When an epoch is finalized it only finalizes the _0th slot_ of the epoch and all prior slots. E.g. if epoch 2 is finalized then slots `<= 64` are finalized, but e.g. slot 65 (which is part of epoch 2) is not.

danielrachi1 replied:
I had no idea 😅
Fixed it (I think) in [this commit](https://github.com/sigp/lighthouse/pull/3753/commits/d4fb8e102643205a66199dab77a0fa89cdb9b723).

michaelsproul commented on common/eth2/src/types.rs:
```diff
@@ -244,6 +244,15 @@ pub struct ExecutionOptimisticForkVersionedResponse<T> {
     pub data: T,
 }
 
+#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
+pub struct ExecutionOptimisticFinalizedForkVersionedResponse<T> {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub version: Option<ForkName>,
+    pub execution_optimistic: Option<bool>,
+    pub finalized: Option<bool>,
+    pub data: T,
+}
```
I think the extra type is OK, because it makes the flags explicit. In future we might be able to consolidate all the `*ForkVersionedResponse` types, although it might come at the cost of some type safety/clarity. I.e. we could use `ExecutionOptimisticFinalizedForkVersionedResponse` for everything and we could set the `finalized` and `execution_optimistic` fields to `Some` where they are required (maybe using a builder pattern like you suggested).

danielrachi1 committed:
Add finalized flag to beacon/blocks/{block_id}/root

michaelsproul reviewed (commented):
This is great, thanks for digging into it!

You're definitely on the right track, I just left a few pointers about how to progress from here

danielrachi1 committed:
fix is_finalized function

danielrachi1 committed:
replace is_finalized (BlockId) for is_finalized_block (BeaconChain)

danielrachi1 committed:
delete is_finalized function

danielrachi1 committed:
Add finalized flag to beacon/blocks/{block_id}/attestations

danielrachi1 committed:
Add finalized flag to beacon/blinded_blocks/{block_id}

danielrachi1 committed:
implement is_finalized_state in BeaconChain

danielrachi1 committed:
Add finalized flag to beacon/states/{state_id}/*

danielrachi1 committed:
Add finalized flag to GET beacon/headers and GET beacon/headers/{block_id}

danielrachi1 committed:
Add finalized flag to GET debug/beacon/states/{state_id}

danielrachi1 commented:
I think that's all :)

Some things to settle before it's... finalized:
- I don't know how I should do the error handling for the two functions I added (`is_finalized_block` and `is_finalized_state`). I don't see how they could fail / when I should return an error. So, any help here would be great. (Maybe just `unwrap_or(false)` ?)
- `execution_optimistic_fork_versioned_response` is no longer used. So, I'll delete it in my next commit. (I wanted to mention this explicitly, just in case.)
- I haven't added any tests. I think I should read the existing tests on `beacon_node/http_api/tests/` and modify them so they check if `finalized` is being returned with the right value. Is this the right way to go?

danielrachi1 committed:
Remove execution_optimistic_fork_versioned_response

danielrachi1 committed:
Proper error handling for the finalized flag

michaelsproul commented:
Sorry I didn't get back to you about this!

Your approach to error handling looks good at a quick glance and I agree we don't need `execution_optimistic_fork_versioned_response` if it's unused.

I'll try to give this a more thorough review on Monday & help with writing tests if you need help

danielrachi1 committed:
Update BeaconNodeHttpClient to handle finalized responses

danielrachi1 committed:
Add tests for the finalized flag

danielrachi commented:
Hmm, It seems like I made some mistakes writing the tests. I couldn't run them properly on my computer, for some reason I'm getting a `linking with cc failed` error.

Besides that, everything should be working now.

danielrachi1 committed:
fix tests

michaelsproul commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -517,11 +517,17 @@ pub fn serve<T: BeaconChainTypes>(
             blocking_json_task(move || {
                 let (root, execution_optimistic) = state_id.root(&chain)?;
-
+                let (state, _) = state_id.state(&chain)?;
+                let state_slot = state.slot();
+                let finalized = chain
+                    .is_finalized_state(&root, state_slot)
+                    .map_err(warp_utils::reject::beacon_chain_error)?;
```
It would be good to avoid loading this `.state()` which is very expensive (potentially multiple minutes for old historic states).

To do this I think we should modify `.root()` so that it also returns finalization information. For each of the cases we can cheaply compute whether the `StateId` is finalized. To enumerate each case:

- `Head`: false
- `Genesis`: true
- `Finalized`: true
- `Justified`: true
- `Root`: true if found in the cold database. If found in the hot database, call `chain.is_finalized_state(root, hot_summary.slot)` (make the `slot` field of `HotStateSummary` public).
- `Slot`: true if slot is less than or equal to `finalized_slot`, false otherwise. We don't need the canonical check (which reads from disk) for `Slot` because it always fetches states from the canonical chain.

danielrachi1 replied:
Shouldn't `Justified` be false? For what I understand from the [docs](https://ethereum.org/en/developers/docs/consensus-mechanisms/pos/#finality), the Justified checkpoint is not finalized but the next one to be finalized.

michaelsproul replied:
Yes! Typo on my part sorry!

michaelsproul commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -548,23 +555,30 @@ pub fn serve<T: BeaconChainTypes>(
+                let ((data, finalized), execution_optimistic) = state_id
+                    .map_state_and_execution_optimistic(&chain, |state, execution_optimistic| {
+                        let state_slot = state.slot();
+                        let (state_root, _) = state_id.root(&chain)?;
```
To avoid re-computing the state root here I think we should modify `map_state_and_execution_optimistic` to also pass the `finalized` status of the state. This will be possible once we've made the change to `.root()` suggested in my first comment and then also modifying `.state()` so that it returns the finalization status.

michaelsproul commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -1094,8 +1165,13 @@ pub fn serve<T: BeaconChainTypes>(
                 };
 
-                Ok(api_types::ExecutionOptimisticResponse {
+                let finalized = chain
+                    .is_finalized_block(&root, block.slot())
+                    .map_err(warp_utils::reject::beacon_chain_error)?;
```
We could probably also return the finalization status from `BlockId::root` using similar logic to the logic for states. This may be worthwhile from a performance point of view because the canonical check will likely form a larger fraction of the total runtime for the block-based APIs (blocks are cheap to load, usually a few milliseconds).

michaelsproul reviewed (changes requested):
This is looking really nice on the whole, I've just made a few suggestions which I hope will decrease the impact on performance.

In general my suggestions try to avoid disk I/O and repetition. Disk I/O, particularly loading full states, is a major source of slowness.

It would also be really cool to do some real-world benchmarks of nodes before and after this change is applied. We could write a utility into `lcli` which uses the API wrappers from `eth2` to make different calls against a live node, and record how long they take. I'd be happy to work on this as it would also be useful for testing other API changes and optimisations.

michaelsproul commented:
I started working on the `http-bench` tool here: https://github.com/michaelsproul/lighthouse/tree/http-bench

It's an `lcli` subcommand currently but I'm going to pull it out into its own repo for more rapid development.

Usage is like this:

```bash
$ RUST_LOG="lcli=debug" lcli http-bench --output results.csv --subjective --num-repeats 3
```

danielrachi1 committed:
StateId::root() now returns info about finalization

danielrachi1 committed:
StateId::map_state_and_execution_optimistic passes the finalized status

danielrachi1 committed:
`BlockId::root` returns info about finalization

danielrachi1 committed:
fix justified finalization response

danielrachi1 committed:
update tests

danielrachi1 commented:
I'm not sure if the `http-bench` tool is ready to use but, just in case, I'm trying to run it and it's panicking.

What I did is:
- Create a branch from the commit before the optimizations [(fix tests)](https://github.com/sigp/lighthouse/pull/3753/commits/5e049e1e3185991fd1aa56c8dbd47507e4ebaf31).
- Add your fork as a remote, pull and merge the `http-bench` branch.
- `cd lighthouse/lcli`
- `RUST_LOG="lcli=debug" cargo run --  http-bench --output results.csv --subjective --num-repeats 3`

This is the output:

```
warning: unreachable pattern
   --> lcli/src/http_bench.rs:245:13
    |
245 |             _ => unimplemented!(),
    |             ^
    |
    = note: `#[warn(unreachable_patterns)]` on by default

warning: `lcli` (bin "lcli") generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 0.30s
     Running `/home/danielrachi/lighthouse/target/debug/lcli http-bench --output results.csv --subjective --num-repeats 3`
thread 'main' panicked at 'Global arguments cannot be required.

        'spec' is marked as global and required', /home/danielrachi/.cargo/registry/src/github.com-1ecc6299db9ec823/clap-2.34.0/src/app/parser.rs:208:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

If you need any help building the tool let me know. I'd be happy to help :)

michaelsproul commented:
> 'spec' is marked as global and required'

@danielrachi I don't get this error when I run it, so I think something must have gotten messed up when you merged `http-bench` into your branch. You could try just building `lcli` from the `http-bench` branch unchanged and running it against different versions of `lighthouse` compiled from your branches. The `lcli` code is independent enough that it won't be affected by the changes to the API implementation (e.g. we could even run it against other non-Lighthouse clients).

I'm reconsidering the approach for the tool, as I think it would be cool to leverage an existing HTTP benchmarking tool like `wrk`.

danielrachi1 commented:
Followed your advice and I'm not getting the error anymore. (And the setup makes more sense with just one executable for each version + lcli instead of the mess I made with the branches, lol.)

However, there was a bigger issue I was ignoring. My SSD has only 512GB. So I'm not able to run a node in order to make the benchmarks 😅

Is there a way of running a node with very limited disk space for this purpose?, if not, I will have to rely on you or someone else to perform the benchmarks.

michaelsproul commented:
> Is there a way of running a node with very limited disk space for this purpose?, if not, I will have to rely on you or someone else to perform the benchmarks.

@danielrachi if you DM me an SSH public key I can give you access to a shared execution node so you can focus on running Lighthouse

EDIT: DM me on Discord (`@sproul#3907`) or send me an email `michael at sigmaprime.io`

danielrachi1 commented:
Using [drill](https://github.com/fcsonline/drill) with [this config](https://github.com/michaelsproul/beacon_http_bench) (skipping `mainnet.yaml`) I got this results:


### Before Optimizations
For commit 5e049e1e3185991fd1aa56c8dbd47507e4ebaf31:
```
$ ITERATIONS=3 drill --benchmark all.yaml --stats

Concurrency 1
Iterations 3
Rampup 0
Base URL http://localhost:5052

head_block_header         http://localhost:5052/eth/v1/beacon/headers/head 200 OK 5ms
head_blinded_block_json   http://localhost:5052/eth/v1/beacon/blinded_blocks/head 200 OK 3ms
head_block_root           http://localhost:5052/eth/v1/beacon/blocks/head/root 200 OK 3ms
head_block_attestations   http://localhost:5052/eth/v1/beacon/blocks/head/attestations 200 OK 3ms
head_state_root           http://localhost:5052/eth/v1/beacon/states/head/root 200 OK 56ms
head_state_fork           http://localhost:5052/eth/v1/beacon/states/head/fork 200 OK 62ms
head_state_finality_checkpoints http://localhost:5052/eth/v1/beacon/states/head/finality_checkpoints 200 OK 0ms
head_blinded_block_ssz    http://localhost:5052/eth/v1/beacon/blinded_blocks/head 200 OK 3ms
head_block_header         http://localhost:5052/eth/v1/beacon/headers/head 200 OK 7ms
head_blinded_block_json   http://localhost:5052/eth/v1/beacon/blinded_blocks/head 200 OK 4ms
head_block_root           http://localhost:5052/eth/v1/beacon/blocks/head/root 200 OK 4ms
head_block_attestations   http://localhost:5052/eth/v1/beacon/blocks/head/attestations 200 OK 3ms
head_state_root           http://localhost:5052/eth/v1/beacon/states/head/root 200 OK 60ms
head_state_fork           http://localhost:5052/eth/v1/beacon/states/head/fork 200 OK 57ms
head_state_finality_checkpoints http://localhost:5052/eth/v1/beacon/states/head/finality_checkpoints 200 OK 0ms
head_blinded_block_ssz    http://localhost:5052/eth/v1/beacon/blinded_blocks/head 200 OK 2ms
head_block_header         http://localhost:5052/eth/v1/beacon/headers/head 200 OK 8ms
head_blinded_block_json   http://localhost:5052/eth/v1/beacon/blinded_blocks/head 200 OK 3ms
head_block_root           http://localhost:5052/eth/v1/beacon/blocks/head/root 200 OK 3ms
head_block_attestations   http://localhost:5052/eth/v1/beacon/blocks/head/attestations 200 OK 2ms
head_state_root           http://localhost:5052/eth/v1/beacon/states/head/root 200 OK 53ms
head_state_fork           http://localhost:5052/eth/v1/beacon/states/head/fork 200 OK 65ms
head_state_finality_checkpoints http://localhost:5052/eth/v1/beacon/states/head/finality_checkpoints 200 OK 0ms
head_blinded_block_ssz    http://localhost:5052/eth/v1/beacon/blinded_blocks/head 200 OK 3ms

head_block_header         Total requests            3
head_block_header         Successful requests       3
head_block_header         Failed requests           0
head_block_header         Median time per request   7ms
head_block_header         Average time per request  6ms
head_block_header         Sample standard deviation 1ms
head_block_header         99.0'th percentile        8ms
head_block_header         99.5'th percentile        8ms
head_block_header         99.9'th percentile        8ms

head_blinded_block_json   Total requests            3
head_blinded_block_json   Successful requests       3
head_blinded_block_json   Failed requests           0
head_blinded_block_json   Median time per request   3ms
head_blinded_block_json   Average time per request  3ms
head_blinded_block_json   Sample standard deviation 1ms
head_blinded_block_json   99.0'th percentile        4ms
head_blinded_block_json   99.5'th percentile        4ms
head_blinded_block_json   99.9'th percentile        4ms

head_block_root           Total requests            3
head_block_root           Successful requests       3
head_block_root           Failed requests           0
head_block_root           Median time per request   3ms
head_block_root           Average time per request  3ms
head_block_root           Sample standard deviation 1ms
head_block_root           99.0'th percentile        4ms
head_block_root           99.5'th percentile        4ms
head_block_root           99.9'th percentile        4ms

head_block_attestations   Total requests            3
head_block_attestations   Successful requests       3
head_block_attestations   Failed requests           0
head_block_attestations   Median time per request   3ms
head_block_attestations   Average time per request  3ms
head_block_attestations   Sample standard deviation 0ms
head_block_attestations   99.0'th percentile        3ms
head_block_attestations   99.5'th percentile        3ms
head_block_attestations   99.9'th percentile        3ms

head_state_root           Total requests            3
head_state_root           Successful requests       3
head_state_root           Failed requests           0
head_state_root           Median time per request   57ms
head_state_root           Average time per request  57ms
head_state_root           Sample standard deviation 3ms
head_state_root           99.0'th percentile        60ms
head_state_root           99.5'th percentile        60ms
head_state_root           99.9'th percentile        60ms

head_state_fork           Total requests            3
head_state_fork           Successful requests       3
head_state_fork           Failed requests           0
head_state_fork           Median time per request   62ms
head_state_fork           Average time per request  61ms
head_state_fork           Sample standard deviation 4ms
head_state_fork           99.0'th percentile        65ms
head_state_fork           99.5'th percentile        65ms
head_state_fork           99.9'th percentile        65ms

head_state_finality_checkpoints Total requests            3
head_state_finality_checkpoints Successful requests       3
head_state_finality_checkpoints Failed requests           0
head_state_finality_checkpoints Median time per request   0ms
head_state_finality_checkpoints Average time per request  0ms
head_state_finality_checkpoints Sample standard deviation 0ms
head_state_finality_checkpoints 99.0'th percentile        0ms
head_state_finality_checkpoints 99.5'th percentile        0ms
head_state_finality_checkpoints 99.9'th percentile        0ms

head_blinded_block_ssz    Total requests            3
head_blinded_block_ssz    Successful requests       3
head_blinded_block_ssz    Failed requests           0
head_blinded_block_ssz    Median time per request   3ms
head_blinded_block_ssz    Average time per request  3ms
head_blinded_block_ssz    Sample standard deviation 0ms
head_blinded_block_ssz    99.0'th percentile        3ms
head_blinded_block_ssz    99.5'th percentile        3ms
head_blinded_block_ssz    99.9'th percentile        3ms

Time taken for tests      0.4 seconds
Total requests            24
Successful requests       24
Failed requests           0
Requests per second       56.52 [#/sec]
Median time per request   3ms
Average time per request  17ms
Sample standard deviation 24ms
99.0'th percentile        65ms
99.5'th percentile        65ms
99.9'th percentile        65ms
```

### After optimizations
For commit 3f07b7e8d123b5c8598453b0c0d54adfc75d7b6d:

```
$ ITERATIONS=3 drill --benchmark all.yaml --stats

Concurrency 1
Iterations 3
Rampup 0
Base URL http://localhost:5052

head_block_header         http://localhost:5052/eth/v1/beacon/headers/head 200 OK 5ms
head_blinded_block_json   http://localhost:5052/eth/v1/beacon/blinded_blocks/head 200 OK 2ms
head_block_root           http://localhost:5052/eth/v1/beacon/blocks/head/root 200 OK 3ms
head_block_attestations   http://localhost:5052/eth/v1/beacon/blocks/head/attestations 200 OK 3ms
head_state_root           http://localhost:5052/eth/v1/beacon/states/head/root 200 OK 0ms
head_state_fork           http://localhost:5052/eth/v1/beacon/states/head/fork 200 OK 49ms
head_state_finality_checkpoints http://localhost:5052/eth/v1/beacon/states/head/finality_checkpoints 200 OK 0ms
head_blinded_block_ssz    http://localhost:5052/eth/v1/beacon/blinded_blocks/head 200 OK 3ms
head_block_header         http://localhost:5052/eth/v1/beacon/headers/head 200 OK 10ms
head_blinded_block_json   http://localhost:5052/eth/v1/beacon/blinded_blocks/head 200 OK 4ms
head_block_root           http://localhost:5052/eth/v1/beacon/blocks/head/root 200 OK 3ms
head_block_attestations   http://localhost:5052/eth/v1/beacon/blocks/head/attestations 200 OK 14ms
head_state_root           http://localhost:5052/eth/v1/beacon/states/head/root 200 OK 0ms
head_state_fork           http://localhost:5052/eth/v1/beacon/states/head/fork 200 OK 70ms
head_state_finality_checkpoints http://localhost:5052/eth/v1/beacon/states/head/finality_checkpoints 200 OK 0ms
head_blinded_block_ssz    http://localhost:5052/eth/v1/beacon/blinded_blocks/head 200 OK 2ms
head_block_header         http://localhost:5052/eth/v1/beacon/headers/head 200 OK 5ms
head_blinded_block_json   http://localhost:5052/eth/v1/beacon/blinded_blocks/head 200 OK 2ms
head_block_root           http://localhost:5052/eth/v1/beacon/blocks/head/root 200 OK 3ms
head_block_attestations   http://localhost:5052/eth/v1/beacon/blocks/head/attestations 200 OK 3ms
head_state_root           http://localhost:5052/eth/v1/beacon/states/head/root 200 OK 0ms
head_state_fork           http://localhost:5052/eth/v1/beacon/states/head/fork 200 OK 49ms
head_state_finality_checkpoints http://localhost:5052/eth/v1/beacon/states/head/finality_checkpoints 200 OK 0ms
head_blinded_block_ssz    http://localhost:5052/eth/v1/beacon/blinded_blocks/head 200 OK 3ms

head_block_header         Total requests            3
head_block_header         Successful requests       3
head_block_header         Failed requests           0
head_block_header         Median time per request   5ms
head_block_header         Average time per request  7ms
head_block_header         Sample standard deviation 2ms
head_block_header         99.0'th percentile        10ms
head_block_header         99.5'th percentile        10ms
head_block_header         99.9'th percentile        10ms

head_blinded_block_json   Total requests            3
head_blinded_block_json   Successful requests       3
head_blinded_block_json   Failed requests           0
head_blinded_block_json   Median time per request   2ms
head_blinded_block_json   Average time per request  3ms
head_blinded_block_json   Sample standard deviation 1ms
head_blinded_block_json   99.0'th percentile        4ms
head_blinded_block_json   99.5'th percentile        4ms
head_blinded_block_json   99.9'th percentile        4ms

head_block_root           Total requests            3
head_block_root           Successful requests       3
head_block_root           Failed requests           0
head_block_root           Median time per request   3ms
head_block_root           Average time per request  3ms
head_block_root           Sample standard deviation 0ms
head_block_root           99.0'th percentile        3ms
head_block_root           99.5'th percentile        3ms
head_block_root           99.9'th percentile        3ms

head_block_attestations   Total requests            3
head_block_attestations   Successful requests       3
head_block_attestations   Failed requests           0
head_block_attestations   Median time per request   3ms
head_block_attestations   Average time per request  7ms
head_block_attestations   Sample standard deviation 5ms
head_block_attestations   99.0'th percentile        14ms
head_block_attestations   99.5'th percentile        14ms
head_block_attestations   99.9'th percentile        14ms

head_state_root           Total requests            3
head_state_root           Successful requests       3
head_state_root           Failed requests           0
head_state_root           Median time per request   0ms
head_state_root           Average time per request  0ms
head_state_root           Sample standard deviation 0ms
head_state_root           99.0'th percentile        0ms
head_state_root           99.5'th percentile        0ms
head_state_root           99.9'th percentile        0ms

head_state_fork           Total requests            3
head_state_fork           Successful requests       3
head_state_fork           Failed requests           0
head_state_fork           Median time per request   49ms
head_state_fork           Average time per request  56ms
head_state_fork           Sample standard deviation 10ms
head_state_fork           99.0'th percentile        70ms
head_state_fork           99.5'th percentile        70ms
head_state_fork           99.9'th percentile        70ms

head_state_finality_checkpoints Total requests            3
head_state_finality_checkpoints Successful requests       3
head_state_finality_checkpoints Failed requests           0
head_state_finality_checkpoints Median time per request   0ms
head_state_finality_checkpoints Average time per request  0ms
head_state_finality_checkpoints Sample standard deviation 0ms
head_state_finality_checkpoints 99.0'th percentile        0ms
head_state_finality_checkpoints 99.5'th percentile        0ms
head_state_finality_checkpoints 99.9'th percentile        0ms

head_blinded_block_ssz    Total requests            3
head_blinded_block_ssz    Successful requests       3
head_blinded_block_ssz    Failed requests           0
head_blinded_block_ssz    Median time per request   3ms
head_blinded_block_ssz    Average time per request  3ms
head_blinded_block_ssz    Sample standard deviation 0ms
head_blinded_block_ssz    99.0'th percentile        3ms
head_blinded_block_ssz    99.5'th percentile        3ms
head_blinded_block_ssz    99.9'th percentile        3ms

Time taken for tests      0.3 seconds
Total requests            24
Successful requests       24
Failed requests           0
Requests per second       95.99 [#/sec]
Median time per request   3ms
Average time per request  10ms
Sample standard deviation 18ms
99.0'th percentile        70ms
99.5'th percentile        70ms
99.9'th percentile        70ms
```

### Comparison 
5e049e1e3185991fd1aa56c8dbd47507e4ebaf31
```
 Time taken for tests      0.4 seconds 
Total requests            24   
Successful requests       24   
Failed requests           0        
Requests per second       56.52 [#/sec]  
Median time per request   3ms   
Average time per request  17ms 
Sample standard deviation 24ms   
99.0'th percentile        65ms   
99.5'th percentile        65ms 
99.9'th percentile        65ms                     
```

3f07b7e8d123b5c8598453b0c0d54adfc75d7b6d
```
Time taken for tests      0.3 seconds
Total requests            24
Successful requests       24
Failed requests           0
Requests per second       95.99 [#/sec]
Median time per request   3ms
Average time per request  10ms
Sample standard deviation 18ms
99.0'th percentile        70ms
99.5'th percentile        70ms
99.9'th percentile        70ms
```
A clear difference in performance is shown. Requests per second almost doubled and all the times improved. So I'd say we can conclude the optimizations you suggested are great.


Just in case, here's the same benchmark for the current version of Lighthouse (`v3.4.0-38514c0`), without the feature implemented:
```
Concurrency 1
Iterations 3
Rampup 0
Base URL http://localhost:5052

head_block_header         http://localhost:5052/eth/v1/beacon/headers/head 200 OK 7ms
head_blinded_block_json   http://localhost:5052/eth/v1/beacon/blinded_blocks/head 200 OK 2ms
head_block_root           http://localhost:5052/eth/v1/beacon/blocks/head/root 200 OK 2ms
head_block_attestations   http://localhost:5052/eth/v1/beacon/blocks/head/attestations 200 OK 2ms
head_state_root           http://localhost:5052/eth/v1/beacon/states/head/root 200 OK 0ms
head_state_fork           http://localhost:5052/eth/v1/beacon/states/head/fork 200 OK 0ms
head_state_finality_checkpoints http://localhost:5052/eth/v1/beacon/states/head/finality_checkpoints 200 OK 0ms
head_blinded_block_ssz    http://localhost:5052/eth/v1/beacon/blinded_blocks/head 200 OK 2ms
head_block_header         http://localhost:5052/eth/v1/beacon/headers/head 200 OK 8ms
head_blinded_block_json   http://localhost:5052/eth/v1/beacon/blinded_blocks/head 200 OK 2ms
head_block_root           http://localhost:5052/eth/v1/beacon/blocks/head/root 200 OK 2ms
head_block_attestations   http://localhost:5052/eth/v1/beacon/blocks/head/attestations 200 OK 2ms
head_state_root           http://localhost:5052/eth/v1/beacon/states/head/root 200 OK 0ms
head_state_fork           http://localhost:5052/eth/v1/beacon/states/head/fork 200 OK 0ms
head_state_finality_checkpoints http://localhost:5052/eth/v1/beacon/states/head/finality_checkpoints 200 OK 0ms
head_blinded_block_ssz    http://localhost:5052/eth/v1/beacon/blinded_blocks/head 200 OK 1ms
head_block_header         http://localhost:5052/eth/v1/beacon/headers/head 200 OK 4ms
head_blinded_block_json   http://localhost:5052/eth/v1/beacon/blinded_blocks/head 200 OK 2ms
head_block_root           http://localhost:5052/eth/v1/beacon/blocks/head/root 200 OK 1ms
head_block_attestations   http://localhost:5052/eth/v1/beacon/blocks/head/attestations 200 OK 2ms
head_state_root           http://localhost:5052/eth/v1/beacon/states/head/root 200 OK 0ms
head_state_fork           http://localhost:5052/eth/v1/beacon/states/head/fork 200 OK 0ms
head_state_finality_checkpoints http://localhost:5052/eth/v1/beacon/states/head/finality_checkpoints 200 OK 0ms
head_blinded_block_ssz    http://localhost:5052/eth/v1/beacon/blinded_blocks/head 200 OK 2ms

head_block_header         Total requests            3
head_block_header         Successful requests       3
head_block_header         Failed requests           0
head_block_header         Median time per request   7ms
head_block_header         Average time per request  6ms
head_block_header         Sample standard deviation 2ms
head_block_header         99.0'th percentile        8ms
head_block_header         99.5'th percentile        8ms
head_block_header         99.9'th percentile        8ms

head_blinded_block_json   Total requests            3
head_blinded_block_json   Successful requests       3
head_blinded_block_json   Failed requests           0
head_blinded_block_json   Median time per request   2ms
head_blinded_block_json   Average time per request  2ms
head_blinded_block_json   Sample standard deviation 0ms
head_blinded_block_json   99.0'th percentile        2ms
head_blinded_block_json   99.5'th percentile        2ms
head_blinded_block_json   99.9'th percentile        2ms

head_block_root           Total requests            3
head_block_root           Successful requests       3
head_block_root           Failed requests           0
head_block_root           Median time per request   2ms
head_block_root           Average time per request  2ms
head_block_root           Sample standard deviation 0ms
head_block_root           99.0'th percentile        2ms
head_block_root           99.5'th percentile        2ms
head_block_root           99.9'th percentile        2ms

head_block_attestations   Total requests            3
head_block_attestations   Successful requests       3
head_block_attestations   Failed requests           0
head_block_attestations   Median time per request   2ms
head_block_attestations   Average time per request  2ms
head_block_attestations   Sample standard deviation 0ms
head_block_attestations   99.0'th percentile        2ms
head_block_attestations   99.5'th percentile        2ms
head_block_attestations   99.9'th percentile        2ms

head_state_root           Total requests            3
head_state_root           Successful requests       3
head_state_root           Failed requests           0
head_state_root           Median time per request   0ms
head_state_root           Average time per request  0ms
head_state_root           Sample standard deviation 0ms
head_state_root           99.0'th percentile        0ms
head_state_root           99.5'th percentile        0ms
head_state_root           99.9'th percentile        0ms

head_state_fork           Total requests            3
head_state_fork           Successful requests       3
head_state_fork           Failed requests           0
head_state_fork           Median time per request   0ms
head_state_fork           Average time per request  0ms
head_state_fork           Sample standard deviation 0ms
head_state_fork           99.0'th percentile        0ms
head_state_fork           99.5'th percentile        0ms
head_state_fork           99.9'th percentile        0ms

head_state_finality_checkpoints Total requests            3
head_state_finality_checkpoints Successful requests       3
head_state_finality_checkpoints Failed requests           0
head_state_finality_checkpoints Median time per request   0ms
head_state_finality_checkpoints Average time per request  0ms
head_state_finality_checkpoints Sample standard deviation 0ms
head_state_finality_checkpoints 99.0'th percentile        0ms
head_state_finality_checkpoints 99.5'th percentile        0ms
head_state_finality_checkpoints 99.9'th percentile        0ms

head_blinded_block_ssz    Total requests            3
head_blinded_block_ssz    Successful requests       3
head_blinded_block_ssz    Failed requests           0
head_blinded_block_ssz    Median time per request   2ms
head_blinded_block_ssz    Average time per request  2ms
head_blinded_block_ssz    Sample standard deviation 0ms
head_blinded_block_ssz    99.0'th percentile        2ms
head_blinded_block_ssz    99.5'th percentile        2ms
head_blinded_block_ssz    99.9'th percentile        2ms

Time taken for tests      0.1 seconds
Total requests            24
Successful requests       24
Failed requests           0
Requests per second       417.42 [#/sec]
Median time per request   2ms
Average time per request  2ms
Sample standard deviation 2ms
99.0'th percentile        8ms
99.5'th percentile        8ms
99.9'th percentile        8ms
```

michaelsproul commented:
nice work on the benchmarks @danielrachi!

the one that stands out is `head_state_fork`, I think the others are all within tolerance (I've seen some of the `head` ones take a while when re-run on the same branch when they just happen to hit lock contention).

from this branch @ 3f07b7e8d123b5c8598453b0c0d54adfc75d7b6d

```
head_state_fork           Total requests            3
head_state_fork           Successful requests       3
head_state_fork           Failed requests           0
head_state_fork           Median time per request   49ms
head_state_fork           Average time per request  56ms
head_state_fork           Sample standard deviation 10ms
head_state_fork           99.0'th percentile        70ms
head_state_fork           99.5'th percentile        70ms
head_state_fork           99.9'th percentile        70ms
```

vs v3.4.0:

```
head_state_fork           Total requests            3
head_state_fork           Successful requests       3
head_state_fork           Failed requests           0
head_state_fork           Median time per request   0ms
head_state_fork           Average time per request  0ms
head_state_fork           Sample standard deviation 0ms
head_state_fork           99.0'th percentile        0ms
head_state_fork           99.5'th percentile        0ms
head_state_fork           99.9'th percentile        0ms
```

I'll see if I can spot the cause for that and suggest a remedy.

michaelsproul commented on beacon_node/http_api/src/state_id.rs:
```diff
@@ -94,35 +112,46 @@ impl StateId {
+        &self,
+        chain: &BeaconChain<T>,
+    ) -> Result<(Fork, bool, bool), warp::Rejection> {
+        let (state, execution_optimistic, finalized) = self.state(chain)?;
```
I think this might be it, lets use `map_state_and_execution_optimistic` instead of `self.state` and see if that improves the `head_state_fork` benchmark.

The reason it's probably better to use the mapping function is that it avoids cloning any beacon states. The head beacon state is at least a few hundred MB in memory and takes on the order of 20-50ms to clone depending on the machine. By using the `map` function we can just pull out the `state.fork()` without cloning the state.

danielrachi1 committed:
Replace .state call for .map_state_and_execution_optimistic

danielrachi1 commented:
@michaelsproul It seems like that was it.

This branch @ b0366ee :
```
head_state_fork           Total requests            3
head_state_fork           Successful requests       3
head_state_fork           Failed requests           0
head_state_fork           Median time per request   0ms
head_state_fork           Average time per request  0ms
head_state_fork           Sample standard deviation 0ms
head_state_fork           99.0'th percentile        1ms
head_state_fork           99.5'th percentile        1ms
head_state_fork           99.9'th percentile        1ms
```

Summary: 
```
Time taken for tests      0.1 seconds
Total requests            24
Successful requests       24
Failed requests           0
Requests per second       200.01 [#/sec]
Median time per request   4ms
Average time per request  4ms
Sample standard deviation 5ms
99.0'th percentile        20ms
99.5'th percentile        20ms
99.9'th percentile        20ms
```

Compared to v3.4.0, summary:
```
Time taken for tests      0.1 seconds
Total requests            24
Successful requests       24
Failed requests           0
Requests per second       280.12 [#/sec]
Median time per request   4ms
Average time per request  3ms
Sample standard deviation 2ms
99.0'th percentile        6ms
99.5'th percentile        6ms
99.9'th percentile        6ms
```

If we want to make it even faster I think we should look at `head_block_header`.
@ b0366ee:
```
head_block_header         Total requests            3
head_block_header         Successful requests       3
head_block_header         Failed requests           0
head_block_header         Median time per request   16ms
head_block_header         Average time per request  15ms
head_block_header         Sample standard deviation 5ms
head_block_header         99.0'th percentile        20ms
head_block_header         99.5'th percentile        20ms
head_block_header         99.9'th percentile        20ms
```

vs v3.4.0

```
head_block_header         Total requests            3
head_block_header         Successful requests       3
head_block_header         Failed requests           0
head_block_header         Median time per request   5ms
head_block_header         Average time per request  5ms
head_block_header         Sample standard deviation 1ms
head_block_header         99.0'th percentile        6ms
head_block_header         99.5'th percentile        6ms
head_block_header         99.9'th percentile        6ms
```

Edit: On that last commit I also fixed a small mistake I made in one of the functions.

danielrachi1 committed:
Fix .map_state_and_execution_optimistic() function name

danielrachi1 committed:
Solve merge conflicts (Merge `unstable` into `add_finalized_flag`)

michaelsproul reviewed (commented):
Looks good

There are a few merge conflicts to fix up but once they're resolved I'll give it another look and we can merge.

Most of the conflicts will be from https://github.com/sigp/lighthouse/pull/4068, which changed all the HTTP handlers to return `warp::reply::Response`. Sprinkling a few `into_response()` calls in a couple of places might be all we need. The `blocking_json_task` already takes care of returning a `Response`, so any calls using that should be OK.

danielrachi1 committed:
Merge branch 'unstable' of github.com:sigp/lighthouse into add_finalized_flag

danielrachi1 commented:
Unless I missed something, it seems like the conflicts were quite simple. Just some ` use` statements that needed to be updated.

michaelsproul commented:
Thanks @danielrachi, seeing as this is quite a major change and isn't essential for Capella I'm nominating it for inclusion in v4.1.0, which should be released in a few weeks.

michaelsproul committed:
Merge remote-tracking branch 'origin/unstable' into add_finalized_flag

michaelsproul committed:
A few clean-ups and optimisations

- Remove unnecessary tuple nesting like `((data, finalized), optimistic)`
- Return `finalized` from `BlockId` functions to avoid repeated block loads from disk
- Add `finalized` status to new rewards APIs added after this PR was opened

michaelsproul commented:
@danielrachi I pushed some minor changes in 5e2be5fee87b73f282d3b5217b1bfd073c01bbb1 that I thought of while reviewing. I'll deploy this PR on some of our nodes, and then I think it's good to merge

michaelsproul committed:
Fix tests

michaelsproul reviewed (approved):
This has been:

- [x] Manually reviewed
- [x] Benchmarked
- [x] Deployed on Goerli

We're good to go

michaelsproul commented:
bors r+

michaelsproul commented:
Will batch this actually

bors r-
bors r+
